### ISSUE TITLE
Problem with input files

### ISSUE THREAD

**Issue by kostasgalexiou** [Biometris/statgenGWAS#7]
Hi,

When I try to generate the gData object, I get the following error:

`> gdata <- createGData(geno = genotypes_transp, map = marker_map, pheno = pheno)`
`Warning messages:`
`1: map contains no marker names. Names constructed from chromosome and position.`
`2: geno contains no genotype names. Names taken from pheno.` 
`3: Not all markers in geno are in map. Extra markers will be removed.`

In spite of the warning I carry on with the vignette to clean the data and I get the following error:

`> gdata_clean = codeMarkers(gData = gdata, verbose = TRUE)`
`Error in codeCharMarkers(markersClean, refAlls, refAll[1] == "minor",  : `
`  Column index is out of bounds: [index=0; column extent=0].`
 
I would appreciate any help. Below are the input files that I am using:

phenotyping file:
[pheno_present_in_genotyping.txt](https://github.com/Biometris/statgenGWAS/files/7425816/pheno_present_in_genotyping.txt)
genotyping file:
[transposed_geno.txt](https://github.com/Biometris/statgenGWAS/files/7425823/transposed_geno.txt)
marker map file:
[geno_map.txt](https://github.com/Biometris/statgenGWAS/files/7425829/geno_map.txt)

I would appreciate any help.

Gracias,
Kostas

**Comment by MARCPUCV** [comment]
Good afternoon everyone, I have a question one realted to generate the Gdata class object to use it later for the coding and imputation of missing data with the codeMarkers () function using Beagle imputation method. There will be an example since I don't know what the format of the files should be that should be generated in Beagle and then generate the gData with that information. Beforehand thank you very much. Kind regards!

**Comment by BartJanvanRossum** [comment]
I am not sure I understand your question. Creating a gData object for imputation with beagle is exactly the same as creating any other gData object. When you call the codeMarkers function and tell it to use Beagle for imputation, the function will take care of converting the format to a format that Beagle can read, then do the imputation and convert the output from Beagle back to the gData format. So there is no need to do anything different when you create your gData object.     
I hope this clarifies things a bit. If not, please let me know

**Comment by MARCPUCV** [comment]
The files with which the Gdata object is formed must be files generated by the Beagle? or can they be .txt files? That part is the one I don't understand. Thank you very much in advance for all your help. Kind regards!!

**Comment by MARCPUCV** [comment]
I get this error and I don't know why? Could it be that the Gdata object is formed with files that are not in Beagle format? or what is this error due to? because the Gdata object has the information of the genotypes and the genetic map and those files are .txt they are not in Beagle format. Thank you very much in advance, best regards !! 
Error in read.table(gzfile(paste0(tmpVcfOut, ".vcf.gz")), stringsAsFactors = FALSE) : 
  no lines available in input
Además: Warning message:
In system(paste0("java -Xmx3000m -jar ", shQuote(paste0(sort(path.package()[grep("statgenGWAS",  :
  comando ejecutado 'java -Xmx3000m -jar "C:/Users/angel/Documents/R/win-library/4.1/statgenGWAS/java/beagle.jar" gt=C:\Users\angel\AppData\Local\Temp\RtmpuMTqcO\file6a5c707611c6input.vcf out=C:\Users\angel\AppData\Local\Temp\RtmpuMTqcO\file6a5c2a363ff6out gp=true seed=1234 nthreads=1 map=C:\Users\angel\AppData\Local\Temp\RtmpuMTqcO\file6a5c3bef6acd.map' tiene estatus 1

**Comment by BartJanvanRossum** [comment]
Hi @MARCPUCV ,
I don't think the way the gData object is formed is the problem. In principle, if you can do imputation with one of the other methods, you should also be able to do imputation using beagle. The codeMarkers function takes care of all the formatting that needs to be done to get you files ready for beagle. 
It is very hard to say what is really going on here without having a look at your actual data and script. Would it be possible to share this with me? A small subset would probably do.

**Comment by MARCPUCV** [comment]
Hi @BartJanvanRossum, 
I send you a small subset of the data sets to your email. Thank you so much for all your help!! Confirm received, please. Kind regards!!

**Comment by BartJanvanRossum** [comment]
I received your files. I will have a look at it

**Comment by BartJanvanRossum** [comment]
I had a look at the files. Your gData object is perfectly fine, no problems there. If I try other types of imputation it is also working.     
The problem is in the fact that beagle requires the position to be in centiMorgan. Judging from your data it seems your positions are in basepairs.     
I just checked the function documentation and the vignette and this isn't mentioned anywhere. I will make sure this gets done in the next version of the package. 
Sorry for the inconvenience.

**Comment by MARCPUCV** [comment]
Hi @BartJanvanRossum,
Thank you very much for all your help, so I must place the information in cM on the genetic map and generate the Gdata with that information to make the imputation with Beagle. I'll tell you how that goes when I do it. Kind regards!!!

**Comment by MARCPUCV** [comment]
Hi @BartJanvanRossum,
I wanted to tell you to put the information in cM but that information is in decimals when I use the codeMarkers function it gives me the following error and I don't know why?
Error in codeMarkers(gData = gData_VAC, impute = TRUE, imputeType = "beagle",  : 
  When using beagle imputation gData should contain a map with only integer positions.

**Comment by BartJanvanRossum** [comment]
Hi @MARCPUCV,
It is more or less exactly what the message says. The beagle software only accepts positions that are in an integer format. Looking at your screenshot multiplying by 100 and then rounding might do the trick for you.

**Comment by MARCPUCV** [comment]
Hi @BartJanvanRossum,
I tried multiplying and rounding round ((100 * G_MAP_VAC $ POS_cM), 0), but when the imputation is done almost all the columns are eliminated which is something that should not happen. What would be happening?

![image](https://user-images.githubusercontent.com/86856783/142200223-b4834d54-def3-40e2-a1bb-9f2dff0628ba.png)

**Comment by MARCPUCV** [comment]
Hi @BartJanvanRossum,
Could you help me with a screenshot of a genetic map that can be used with the Beagle imputation for reference please.
Best regards!!

**Comment by BartJanvanRossum** [comment]
After some digging done (see #8) it seems that something (not yet sure what) changed in the input format for beagle 5.2 compared to the previous version used in the package (beagle 4.1). For now I fixed this by putting back beagle 4.1. It also seems to solve your problem with the imputation.    
You can use your original input, no need to divide by 1000. 

You can get the latest version by installing the package from the develop branch:
```r
remotes::install_github("Biometris/statgenGWAS", ref = "develop", dependencies = TRUE)
```

I will still try to figure out why beagle 5.2 doesn't work, but at the very least you can do the imputation now.

**Comment by MARCPUCV** [comment]
Hi @BartJanvanRossum,
I tell you that I did what you told me in the previous message and the imputation with Beagle worked perfectly considering the bp position on the genetic map. Thank you very much and I managed to make the imputation with Beagle with your indications from the last message. Very grateful for your help. Kind regards,
Angelica

**Comment by BartJanvanRossum** [comment]
Closing this as the original issue is solved and the beagle issue tracked in #8
